File: backend/app/schemas/training.py
"""Training job schemas."""

from datetime import datetime
from typing import Optional
from uuid import UUID
from pydantic import BaseModel, Field

from app.models.training import TrainingStatus, FineTuningMethod


class TrainingJobBase(BaseModel):
    """Base training job schema."""

    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None


class TrainingJobCreate(TrainingJobBase):
    """Schema for creating a training job."""

    project_id: UUID
    base_model_id: UUID
    dataset_id: UUID
    fine_tuning_method: FineTuningMethod
    hyperparameters: Optional[dict] = None


class TrainingJobUpdate(BaseModel):
    """Schema for updating a training job."""

    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None


class TrainingJobResponse(TrainingJobBase):
    """Schema for training job response."""

    id: UUID
    project_id: UUID
    base_model_id: UUID
    dataset_id: UUID
    fine_tuning_method: FineTuningMethod
    hyperparameters: dict
    status: TrainingStatus
    progress: float
    current_epoch: int
    total_epochs: int
    current_step: int
    total_steps: Optional[int]
    metrics: dict
    latest_train_loss: Optional[float]
    latest_eval_loss: Optional[float]
    best_eval_loss: Optional[float]
    logs: str
    error_message: Optional[str]
    output_model_path: Optional[str]
    output_model_id: Optional[UUID]
    compute_used: dict
    checkpoint_dir: Optional[str]
    best_checkpoint_path: Optional[str]
    created_by: UUID
    created_at: datetime
    updated_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]

    class Config:
        from_attributes = True


class TrainingConfigResponse(BaseModel):
    """Training configuration response."""

    id: UUID
    name: str
    description: Optional[str]
    model_type: str
    method: FineTuningMethod
    default_hyperparameters: dict
    is_public: bool
    organization_id: Optional[UUID]
    created_by: Optional[UUID]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/training.ts
export enum TrainingStatus {
  QUEUED = 'queued',
  INITIALIZING = 'initializing',
  RUNNING = 'running',
  PAUSED = 'paused',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

export enum FineTuningMethod {
  LORA = 'lora',
  QLORA = 'qlora',
  FULL = 'full',
  PEFT = 'peft',
}

export interface TrainingJob {
  id: string;
  name: string;
  description?: string;
  project_id: string;
  base_model_id: string;
  dataset_id: string;
  fine_tuning_method: FineTuningMethod;
  hyperparameters: Record<string, any>;
  status: TrainingStatus;
  progress: number;
  current_epoch: number;
  total_epochs: number;
  current_step: number;
  total_steps?: number;
  metrics: Record<string, any>;
  latest_train_loss?: number;
  latest_eval_loss?: number;
  best_eval_loss?: number;
  logs: string;
  error_message?: string;
  output_model_path?: string;
  output_model_id?: string;
  compute_used: Record<string, any>;
  checkpoint_dir?: string;
  best_checkpoint_path?: string;
  created_by: string;
  created_at: string;
  updated_at: string;
  started_at?: string;
  completed_at?: string;
}

export interface TrainingConfig {
  id: string;
  name: string;
  description?: string;
  model_type: string;
  method: FineTuningMethod;
  default_hyperparameters: Record<string, any>;
  is_public: boolean;
  organization_id?: string;
  created_by?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateTrainingJobData {
  name: string;
  description?: string;
  project_id: string;
  base_model_id: string;
  dataset_id: string;
  fine_tuning_method: FineTuningMethod;
  hyperparameters?: Record<string, any>;
}
